mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,459 @@
+                    " to specification in the config file."
+                )
+                field.attrs[attribute] = field_attributes[attribute]
+            else:
+                # Attributes are set and nothing has been defined in the config file
+                pass
+
+    return field
+
+
+def _return_dropped_coordinates(ds_subset, ds_input, required_coordinates, chunks):
+    """Return the coordinates that have been reset."""
+    for req_coord in required_coordinates:
+        if req_coord in chunks:
+            ds_subset.coords[req_coord] = ds_input[req_coord]
+
+    return ds_subset
+
+
+def calculate_toa_radiation(lat, lon, time):
+    """
+    Function for calculating top-of-the-atmosphere radiation
+
+    Parameters
+    ----------
+    lat : xr.DataArray or float
+        Latitude values
+    lon : xr.DataArray or float
+        Longitude values
+    time : xr.DataArray or datetime object
+        Time
+
+    Returns
+    -------
+    toa_radiation: xr.DataArray or float
+        TOA radiation data
+    """
+    logger.info("Calculating top-of-atmosphere radiation")
+
+    # Solar constant
+    E0 = 1366  # W*m**-2
+
+    day = time.dt.dayofyear
+    hr_utc = time.dt.hour
+
+    # Eq. 1.6.1a in Solar Engineering of Thermal Processes 4th ed.
+    dec = np.pi / 180 * 23.45 * np.sin(2 * np.pi * (284 + day) / 365)
+
+    hr_lst = hr_utc + lon / 15
+    hr_angle = 15 * (hr_lst - 12)
+
+    # Eq. 1.6.2 with beta=0 in Solar Engineering of Thermal Processes 4th ed.
+    cos_sza = np.sin(lat * np.pi / 180) * np.sin(dec) + np.cos(
+        lat * np.pi / 180
+    ) * np.cos(dec) * np.cos(hr_angle * np.pi / 180)
+
+    # Where TOA radiation is negative, set to 0
+    toa_radiation = xr.where(E0 * cos_sza < 0, 0, E0 * cos_sza)
+
+    if isinstance(toa_radiation, xr.DataArray):
+        # Add attributes
+        toa_radiation.name = "toa_radiation"
+        toa_radiation.attrs["long_name"] = "top-of-the-atmosphere radiation"
+        toa_radiation.attrs["units"] = "W*m**-2"
+
+    return toa_radiation
+
+
+def calculate_hour_of_day(time):
+    """
+    Function for calculating hour of day features with a cyclic encoding
+
+    Parameters
+    ----------
+    time : xr.DataArray or datetime object
+        Time
+
+    Returns
+    -------
+    hour_of_day_cos: xr.DataArray or float
+        cosine of the hour of day
+    hour_of_day_sin: xr.DataArray or float
+        sine of the hour of day
+    """
+    logger.info("Calculating hour of day")
+
+    # Get the hour of the day
+    hour_of_day = time.dt.hour
+
+    # Cyclic encoding of hour of day
+    hour_of_day_cos, hour_of_day_sin = cyclic_encoding(hour_of_day, 24)
+
+    if isinstance(hour_of_day_cos, xr.DataArray):
+        # Add attributes
+        hour_of_day_cos.name = "hour_of_day_cos"
+        hour_of_day_cos.attrs[
+            "long_name"
+        ] = "Cosine component of cyclically encoded hour of day"
+        hour_of_day_cos.attrs["units"] = "1"
```

Comment:
"1" is used to indicate no unit? Would it be better with None or empty string?

Reply:
According to [CF conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch03.html)
> The conforming unit for quantities that represent fractions, or parts of a whole, is "1"

I would guess that this applies for sine and cosine quantities?

Reply:
"1" is common for dimensionless variables. In fact this doesn't mean "does not have a unit" but the unit is dimensionless :) We should keep "1" since that that is also the CF compliant way of indicating that a variable is dimensionless (http://cfconventions.org/cf-conventions/cf-conventions.html#dimensionless-units)